admin/billing/orb.go

```diff
@@ -0,0 +1,378 @@
+			StartDate:               s.StartDate,
+			EndDate:                 s.EndDate,
+			CurrentBillingStartDate: s.CurrentBillingPeriodStartDate,
+			CurrentBillingEndDate:   s.CurrentBillingPeriodEndDate,
+			TrialEndDate:            s.TrialInfo.EndDate,
+			Metadata:                s.Metadata,
+		})
+	}
+	return subscriptions, nil
+}
+
+func (o *Orb) CancelSubscription(ctx context.Context, subscriptionID string, cancelOption SubscriptionCancellationOption, cancellationDate time.Time) error {
+	var cancelParams orb.SubscriptionCancelParams
+	switch cancelOption {
+	case SubscriptionCancellationOptionEndOfSubscriptionTerm:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption: orb.F(orb.SubscriptionCancelParamsCancelOptionEndOfSubscriptionTerm),
+		}
+	case SubscriptionCancellationOptionImmediate:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption: orb.F(orb.SubscriptionCancelParamsCancelOptionImmediate),
+		}
+	case SubscriptionCancellationOptionRequestedDate:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption:     orb.F(orb.SubscriptionCancelParamsCancelOptionRequestedDate),
+			CancellationDate: orb.F(cancellationDate),
+		}
+	}
+
+	_, err := o.client.Subscriptions.Cancel(ctx, subscriptionID, cancelParams)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *Orb) CancelSubscriptionsForCustomer(ctx context.Context, customerID string, cancelOption SubscriptionCancellationOption, cancellationDate time.Time) error {
+	var cancelParams orb.SubscriptionCancelParams
+	switch cancelOption {
+	case SubscriptionCancellationOptionEndOfSubscriptionTerm:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption: orb.F(orb.SubscriptionCancelParamsCancelOptionEndOfSubscriptionTerm),
+		}
+	case SubscriptionCancellationOptionImmediate:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption: orb.F(orb.SubscriptionCancelParamsCancelOptionImmediate),
+		}
+	case SubscriptionCancellationOptionRequestedDate:
+		cancelParams = orb.SubscriptionCancelParams{
+			CancelOption:     orb.F(orb.SubscriptionCancelParamsCancelOptionRequestedDate),
+			CancellationDate: orb.F(cancellationDate),
+		}
+	}
+
+	subs, err := o.GetSubscriptionsForCustomer(ctx, customerID)
+	if err != nil {
+		return err
+	}
+	for _, s := range subs {
+		_, err := o.client.Subscriptions.Cancel(ctx, s.ID, cancelParams)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *Orb) ReportUsage(ctx context.Context, customerID string, usage []*Usage) error {
+	var orbUsage []orb.EventIngestParamsEvent
+	for _, u := range usage {
+		eventName := u.MetricName + "_" + string(u.ReportingGran)
+		// use end time minus 1 second to make sure the event is attributed to the current time bucket
+		eventTime := u.EndTime.Add(-1 * time.Second)
+		// generate idempotency key using customer id, timestamp, event name and metadata
+		idempotencyKey := fmt.Sprintf("%s_%d_%s_%v", customerID, eventTime.UnixMilli(), eventName, u.Metadata)
```

Comment:
but the final usage event might not have 1:1 mapping with event_id because the api may may do some aggregation, key is used to make sure duplicate usage is not reported for an interval.